### PR TITLE
feat: proxy-mirror support grpc and grpcs

### PR DESCRIPTION
**Pull request by Sn0rt**
### Description

Fixes https://github.com/apache/apisix/issues/7437 https://github.com/apache/apisix/issues/9245

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
No test cases?

**Comment by Sn0rt** [comment]
> No test cases?

sorry. The PR is POC. I push a new PR by force. thx for you review.

**Comment by Sn0rt** [comment]
@monkeyDluffy6017 pls help me to re-try the ci.

File: apisix/init.lua [hunk 1]
```diff
@@ -688,6 +689,13 @@ function _M.grpc_access_phase()
         core.log.error("failed to set grpcs upstream param: ", err)
         core.response.exit(code)
     end
+
+    if api_ctx.var.upstream_mirror_scheme == "grpc" or
+        api_ctx.var.upstream_mirror_scheme == "grpcs" then
+        if has_mod then
+            apisix_ngx_client.enable_mirror()
+        end
+    end
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:698]: Why do you enable mirror here? it should be set in `proxy-mirror` plugin

Comment (Sn0rt) [reply apisix/init.lua:698]: You are right, if you want to support the plug-in list like http, the code may look like the following.

```diff
diff --git a/apisix/init.lua b/apisix/init.lua
index 56228670..f4ed438c 100644
--- a/apisix/init.lua
+++ b/apisix/init.lua
@@ -690,12 +690,20 @@ function _M.grpc_access_phase()
         core.response.exit(code)
     end

-    if api_ctx.var.upstream_mirror_scheme == "grpc" or
-        api_ctx.var.upstream_mirror_scheme == "grpcs" then
-        if has_mod then
-            apisix_ngx_client.enable_mirror()
-        end
+    local route = api_ctx.matched_route
+    if not route then
+        core.log.info("not find any matched route")
+        return core.response.exit(404,
+                    {error_msg = "404 Route Not Found"})
     end
+
+    core.log.info("matched route: ",
+            core.json.delay_encode(api_ctx.matched_route, true))
+
+    local plugins = plugin.filter(api_ctx, route)
+    api_ctx.plugins = plugins
+
+    plugin.run_plugin("rewrite", plugins, api_ctx)
 end

```

Comment (monkeyDluffy6017) [reply apisix/init.lua:698]: I think removing this code is enough

**Comment by Sn0rt** [comment]
@monkeyDluffy6017 pls help to re-try the CI

File: apisix/init.lua [hunk 2]
```diff
@@ -711,11 +712,16 @@ function _M.grpc_access_phase()
         return
     end
 
+
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:715]: remove this blank

File: apisix/plugins/proxy-mirror.lua [hunk 3]
```diff
@@ -114,12 +117,14 @@ function _M.rewrite(conf, ctx)
 
     if conf.sample_ratio == 1 then
         enable_mirror(ctx, conf)
+        ctx.enable_mirror = 1
```
Comment (monkeyDluffy6017) [inline apisix/plugins/proxy-mirror.lua:120]: Is `true` better?

File: t/APISIX.pm [hunk 4]
```diff
@@ -743,6 +744,8 @@ _EOC_
         }
 
         location / {
+            set \$upstream_mirror_scheme      '';
```
Comment (monkeyDluffy6017) [inline t/APISIX.pm:747]: upstream_mirror_scheme  is not used?

File: docs/zh/latest/plugins/proxy-mirror.md [hunk 5]
```diff
@@ -40,9 +40,9 @@ description: 本文介绍了 Apache APISIX proxy-mirror 插件的相关操作，
 
 | 名称 | 类型   | 必选项 | 默认值 | 有效值 | 描述                                                                                                    |
 | ---- | ------ | ------ | ------ | ------ | ------------------------------------------------------------------------------------------------------- |
-| host | string | 是   |        |        | 指定镜像服务的地址，地址中需要包含 `schema`（`http` 或 `https`），但不能包含 `path` 部分。例如 `http://127.0.0.1:9797`。 |
-| path | string | 否   |        |        | 指定镜像请求的路径。如果不指定，则默认会使用当前路径。 |
-| path_concat_mode | string | 否   |   replace     | ["replace", "prefix"]       | 当指定镜像请求的路径时，设置请求路径的拼接模式。`replace` 模式将会直接使用 `path` 作为镜像请求的路径。`prefix` 模式将会使用 `path` + `来源请求 URI` 作为镜像请求的路径。  |
+| host | string | 是   |        |        | 指定镜像服务的地址，地址中需要包含 `schema`（`http(s)` 或 `grpc(s)`），但不能包含 `path` 部分。例如 `http://127.0.0.1:9797`。 |
+| path | string | 否   |        |        | 指定镜像请求的路径。如果不指定，则默认会使用当前路径。如果是为了镜像 grpc 流量，这个选项不再适用。|
```
Comment (monkeyDluffy6017) [inline docs/zh/latest/plugins/proxy-mirror.md:44]: could you give an example about how to use grpc mirror in doc?

Comment (Sn0rt) [reply docs/zh/latest/plugins/proxy-mirror.md:44]: At present, the document format is a case

File: apisix/init.lua (hunk 2 above)
Comment (Sn0rt) [reply apisix/init.lua:715]: done

File: apisix/core/ctx.lua [hunk 6]
```diff
@@ -190,6 +190,8 @@ do
         upstream_connection        = true,
         upstream_uri               = true,
 
+        upstream_mirror_scheme     = true,
```
Comment (monkeyDluffy6017) [inline apisix/core/ctx.lua:193]: Why not delete this

Comment (Sn0rt) [reply apisix/core/ctx.lua]: loss. now removed

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**